Programming Question Audit Trail

syoopie opened the pull request:
## Motivation

**Old Behavior** - When question is updated and regrading is triggered, there was no preservation of the:
1. Previous question state
2. Previous test case evaluations

This resulted in a dilemma when randomness in test cases (such as in machine learning) resulted in a student failing test cases that were previously passing


## Details

<img width="1453" alt="image" src="https://github.com/user-attachments/assets/65f6d925-fe71-482e-8a44-a6d6275e1208">

- parentId added to `course_assessment_question_programming` and `course_assessment_answer_programming_auto_gradings`
- All evaluation data is preserved now
- Old evaluations and there corresponding questions can be accessed via the parentId

syoopie committed:
feat(programming-audit-trail): modify db

add parentId for:
- programming auto grading
- programming question

syoopie committed:
feat(custom-slider): add custom slider

syoopie committed:
feat(accordion): expand accordion functionality

- include an option to add elipses beside the title to better indicate that accordion can be opened

syoopie committed:
refactor(statistics): change naming of types

- QuestionDetails -> Question
- AllAnswerDetails -> Answer

- previous naming was verbose and communicated the wrong idea of what information the type has
- free up naming for use in details components

syoopie committed:
refactor(all-attempts-display): use new slider

syoopie committed:
refactor(all-attempts-display): update accordion

- make new of new accordion ellipses feature

syoopie committed:
refactor(all-attempts-display): change created_at to submitted_at

- submitted_at is more indicative of the actual data

syoopie committed:
feat(statistics): add BE support for multiple tests / questions

syoopie committed:
feat(statistics): multiple regrading view for attempts

- updated types for processed vs non processed data
- removed unnecessary data from jbuilders
- added parsing / processing of programming question data and linked to display

syoopie committed:
feat(regrading): only regrade current_answer on question edit

syoopie committed:
feat(all-attempts-display): add more programming question details

syoopie committed:
test(stats-answer-controller): rename function calls

- lines removed is data that is no longer being rendered

syoopie committed:
test(programming-controller): fix tests

- use update column instead of update to prevent validation error when duplicating question with deprecated programming language

syoopie committed:
test(programming-management): fix tests

- also add redicrect edit url to return when updating question, so that id can be used to revalidate question

syoopie committed:
test(programming-management): fix template package spec

syoopie committed:
fix(attempt-count-table): fix error when attempt does not exist

- this situation can occur when question is added after submission

syoopie committed:
test(test-case-view): fix FE test

cysjonathan commented on app/controllers/course/statistics/answers_controller.rb:
```diff
@@ -33,7 +33,7 @@ def all_answers
     @all_answers = Course::Assessment::Answer.
                    unscope(:order).
-                   order(:created_at).
+                   order(:submitted_at).
```
Since we also want to display submissions in attempting state, what happens to answers with `nil` for `submitted_at`?

syoopie replied:
hm I dont think there is a way to reconcile that. But do you think its ok for us to order by `created_at` and also send the `created_at` time but present it to the user as `submitted_at` (which is what was done previously

cysjonathan replied:
It's ok, since user has not run the code, it will not be counted as an attempt, so it will reflect as 0 on attempts page, (or whatever number depending on how many past attempts they already had existing).

syoopie replied:
Ok changed to order by created_at

cysjonathan commented on app/controllers/course/statistics/answers_controller.rb:
```diff
@@ -4,6 +4,24 @@ class Course::Statistics::AnswersController < Course::Statistics::Controller
 
   MAX_ANSWERS_COUNT = 10
 
+  def latest_attempt
```
I feel this should be renamed as `latest_answer` to communicate that we are retrieving an answer.

syoopie replied:
Done

cysjonathan commented on app/controllers/course/statistics/answers_controller.rb:
```diff
@@ -22,7 +22,7 @@ def latest_attempt
   end
 
-  def question_answer_details
+  def attempts
```
similarly why is this `attempts` not `answers`?

syoopie replied:
Attempt aligns with the vocab used in frontend, since we have never exposed the concept of "answer" to users

cysjonathan replied:
Sure, do leave a comment that an "attempt" is an answer that has been submitted before.

syoopie replied:
Done

cysjonathan commented on app/controllers/course/statistics/answers_controller.rb:
```diff
@@ -76,17 +76,17 @@ def question_index(question_id)
 
   def fetch_all_answers(submission_id, question_id, limit)
-    @all_answers = if limit
+    @all_answers = if limit == -1
```
Ok I see you've updated the query to adjust the `limit` parameter to take in a number. Please squash the relevant commits.

syoopie replied:
Done

cysjonathan commented on app/controllers/course/statistics/answers_controller.rb:
```diff
@@ -39,10 +39,10 @@ def question_answer_details
   end
 
-  def all_answers
+  def all_attempts
```
Similar comment, `all_answers` vs `all_attempts`

syoopie replied:
Attempt aligns with the vocab used in frontend, since we have never exposed the concept of "answer" to users

cysjonathan commented on app/controllers/course/assessment/question/programming_controller.rb:
```diff
@@ -44,8 +44,8 @@ def edit
   def update
-    duplicate_and_replace_programming_question
     result = @programming_question.class.transaction do
+      duplicate_and_replace_programming_question
```
If you're rearranging items, please squash the commit with the respective commit?

syoopie replied:
Done

syoopie commented on app/views/course/statistics/answers/all_answers.json.jbuilder:
```diff
@@ -3,6 +3,11 @@ is_displayed = @submission.graded? || @submission.published?
 json.isAnswersDisplayed is_displayed
 
+json.user do
+  json.name @submission.creator.name.upcase
```
I think during implementation I thought it looked better or something. 

Removed usage of `.upcase` since it can be done in FE if need be

syoopie commented on app/views/course/assessment/question/programming/_programming.json.jbuilder:
```diff
@@ -1,7 +1,13 @@
-json.isCodaveri question.is_codaveri
-json.liveFeedbackEnabled question.live_feedback_enabled
+
+json.isCodaveri question.is_codaveri?
+json.liveFeedbackEnabled question.live_feedback_enabled?
```
Typically within the codebase and in ruby booleans are all decorated with `?`
So i changed it as such to indicate that it is a boolean

syoopie commented on app/controllers/course/statistics/answers_controller.rb:
```diff
@@ -56,12 +57,18 @@ def question_index(question_id)
-              where(submission_id: submission_id, question_id: question_id)
-
-    @all_answers = answers.limit(MAX_ANSWERS_COUNT)
+  def fetch_all_answers(submission_id, question_id, limit)
```
Done

syoopie committed:
feat(programming-audit-trail): preserve old questions and tests

syoopie committed:
refactor(statistics): use seperate API for latest answer

- changed naming for files and added specific typing for latest answer
- fix bug where if a question is added after submission, the page crashes

syoopie committed:
feat(statistics-marks-table): add card to dialog

bivanalhar committed:
refactor(pastAnswer): follow audit trail
